include/revng/Pipeline/Kind.h

```diff
@@ -0,0 +1,76 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+//
+#include <set>
+
+#include "revng/Pipeline/DynamicHierarchy.h"
+#include "revng/Pipeline/InvalidationEvent.h"
+#include "revng/Pipeline/Quantifier.h"
+#include "revng/Support/Assert.h"
+
+namespace Pipeline {
+class Target;
+class Context;
+
+/// The granularity tree is a tree used by targets to
+/// find out how many names are required to name a target.
+class Granularity : public DynamicHierarchy<Granularity> {
+private:
+  static std::vector<Granularity *> Roots;
+
+public:
+  friend DynamicHierarchy;
+
+  Granularity(llvm::StringRef Name) : DynamicHierarchy(Name) {}
+  Granularity(llvm::StringRef Name, Granularity &Parent) :
+    DynamicHierarchy<Granularity>(Name, Parent) {}
+};
+
+class TargetsList;
+
+class Kind : public DynamicHierarchy<Kind> {
+private:
+  static std::vector<Kind *> Roots;
+
+public:
+  friend DynamicHierarchy;
+
+  Kind(llvm::StringRef Name, Granularity *Granularity) :
+    DynamicHierarchy<Kind>(Name), Granularity(Granularity) {
+    revng_assert(Granularity != nullptr);
+  }
+  Kind(llvm::StringRef Name, Kind &Parent, Granularity *Granularity) :
+    DynamicHierarchy<Kind>(Name, Parent), Granularity(Granularity) {
+    revng_assert(Granularity != nullptr);
+  }
+
+  virtual void deduceInvalidations(const InvalidationEventBase &,
+                                   std::set<GranularityList> &Targets) const {
```

Comment:
Do we ever want users to call this? If not, make it private and rename `recursiveDeduceInvalidations` in `deduceInvalidations`.

Reply:
but we should use it some day

Reply:
Yes, but I think `deduceInvalidations` is an implementation detail that should be private.
Users will always want to call `recursiveDeduceInvalidations`, am I right?